### PR TITLE
Fix spurious unlock in close_db_if_idle

### PR DESCRIPTION
**Pull request by rnewson**
We've observed `couch_db:incref/1` being called with `undefined` under load. On examination of couch_server it seems we unlock an `#entry` unilaterally, which can leave an `#entry` where lock is `unlocked` but `db` is not yet populated, and hence the value based to `incref` is the default for the record, which is `undefined`.

This patch modifies close_db_if_idle to only proceed if it finds an unlocked #entry.

### CODE REVIEW

File: src/couch/src/couch_lru.erl [hunk 1]
```diff
@@ -46,9 +46,8 @@ close_int({Lru, DbName, Iter}, {Tree, Dict} = Cache) ->
     CouchDbs = couch_server:couch_dbs(DbName),
     CouchDbsPidToName = couch_server:couch_dbs_pid_to_name(DbName),
 
-    case ets:update_element(CouchDbs, DbName, {#entry.lock, locked}) of
-        true ->
-            [#entry{db = Db, pid = Pid}] = ets:lookup(CouchDbs, DbName),
+    case couch_server:try_lock(DbName, CouchDbs) of
```
Comment (nickva) [inline src/couch/src/couch_lru.erl:49]: Argument order is reversed here CouchDbs is the table reference and DbName should be key.

File: src/couch/src/couch_server.erl [hunk 2]
```diff
@@ -1011,6 +1011,21 @@ names() ->
     N = couch_server:num_servers(),
     [couch_server:couch_server(I) || I <- lists:seq(1, N)].
 
+%% Try to lock an entry, must be unlocked at the time.
+try_lock(Table, DbName) ->
```
Comment (nickva) [inline src/couch/src/couch_server.erl:1015]: Might be a good idea to assert the types, that `Table` is an atom and `DbName` a binary?

File: src/couch/src/couch_server.erl [hunk 3]
```diff
@@ -31,6 +31,7 @@
 -export([num_servers/0, couch_server/1, couch_dbs_pid_to_name/1, couch_dbs/1]).
 -export([aggregate_queue_len/0, get_spidermonkey_version/0]).
 -export([names/0]).
+-export([try_lock/2]).
```
Comment (nickva) [inline src/couch/src/couch_server.erl:34]: Should we have a symmetric `unlock/2` function as well. Just so locking/unlocking happens in those utility functions only?

**Review by nickva — CHANGES_REQUESTED** [review]
Nice use of CAS select_replace feature!

Just noticed the arguments were reversed in the lru and had a few minor suggestions.

File: src/couch/src/couch_server.erl (hunk 3 above)
Comment (rnewson) [reply src/couch/src/couch_server.erl:34]: I thought about it but it wouldn't be "symmetric" (in the sense that try_unlock would only unlock if locked, using another select_replace). We _know_ we acquired the lock, so we can release it without checking. Is that worth pulling out into an `unlock` function which is just the `update_element` function? I could do it, not sure it clarifies much. thoughts?

File: src/couch/src/couch_lru.erl (hunk 1 above)
Comment (rnewson) [reply src/couch/src/couch_lru.erl]: yeah..

File: src/couch/src/couch_server.erl (hunk 2 above)
Comment (rnewson) [reply src/couch/src/couch_server.erl]: have done this, good thought.

File: src/couch/src/couch_server.erl (hunk 3 above)
Comment (rnewson) [reply src/couch/src/couch_server.erl:34]: sure.

**Review by nickva — APPROVED** [review]
+1

**Pull request merged**